river/cluster/clustream.py

```diff
@@ -0,0 +1,363 @@
+                        M=self.max_kernels,
+                    )
+            self.buffer.clear()
+            self.initialized = True
+
+            return self
+
+        """determine closest kernel"""
+        closest_kernel = None
+        min_distance = sys.float_info.max
+        for kernel in self.kernels.values():
+            distance = self._distance(x, kernel.center)
+            if distance < min_distance:
+                closest_kernel = kernel
+                min_distance = distance
+
+        """check whether the instance fits into closest kernel"""
+        radius = 0.0
+        if closest_kernel.weight == 1:
+            radius = sys.float_info.max
+            center = closest_kernel.center
+            for kernel in self.kernels.values():
+                if kernel == closest_kernel:
+                    continue
+                distance = self._distance(kernel.center, center)
+                radius = min(distance, radius)
+        else:
+            radius = closest_kernel.radius
+
+        if min_distance < radius:
+            closest_kernel.insert(x, sample_weight, self.time_stamp)
+            return self
+
+        """Data does not fit, we need to free some space in order to insert a new kernel"""
+
+        threshold = self.time_stamp - self.time_window
+
+        """try to delete old kernel"""
+        for i, kernel in self.kernels.items():
+            if kernel.relevance_stamp < threshold:
+                self.kernels[i] = ClustreamKernel(
+                    x=x,
+                    sample_weight=sample_weight,
+                    timestamp=self.time_stamp,
+                    T=self.kernel_radius_factor,
+                    M=self.max_kernels,
+                )
+
+                return self
+
+        """try to merge closest two kernels"""
+        closest_a = 0
+        closest_b = 0
+        min_distance = sys.float_info.max
+        for i, kernel_i in self.kernels.items():
+            center_a = kernel_i.center
+            for j, kernel_j in self.kernels.items():
+                dist = self._distance(center_a, kernel_j.center)
+                if dist < min_distance and j > i:
+                    min_distance = dist
+                    closest_a = i
+                    closest_b = j
+        self.kernels[closest_a].add(self.kernels[closest_b])
+        self.kernels[closest_b] = ClustreamKernel(
+            x=x,
+            sample_weight=sample_weight,
+            timestamp=self.time_stamp,
+            T=self.kernel_radius_factor,
+            M=self.max_kernels,
+        )
+
+        return self
+
+    def get_micro_clustering_result(self):
+
+        if not self.initialized:
+            return {}
+        res = {
+            i: ClustreamKernel(
+                cluster=self.kernels[i], T=self.kernel_radius_factor, M=self.max_kernels
+            )
+            for i in range(len(self.kernels))
+        }
+        return res
+
+    def learn_predict_one(self, x, sample_weight=None):
```

Comment:
I wouldn't add this function. Fundamentally it doesn't do anything more than calling `learn_one` followed by `predict_one`.

Reply:
Since `learn_one` returns `self`, the user can do `clusterer.learn_one(x).predict_one(x)`. I guess it is kind of elegant to have a single call, but have no strong opinion. 

However, I see two issues:
- Should the order be inverted? `predict_learn(x)`
- If we keep this method, we should make it a method of the `Clusterer` class.

Comments @MaxHalford @smastelini ?

Reply:
I would keep something of this nature only if the speedups were clear. As @MaxHalford said, the function chains the call to two other already available functions.

Regarding the ordering, I am not sure hahaha
In general, clustering is a whole different monster when compared to supervised learning. The task is descriptive rather than predictive. So, I am really not sure we need to "predict" before learning.

Reply:
Thanks for the comments to both. We agree that the functionality can be achieved by calling `learn_one` and `predict_one`. On the other hand, the order if we were to keep this function is not really clear. So let's remove it for the moment and only if a use-case is identified we would re-evaluate its implementation.

Reply:
Thanks all for all of your comments! After the discussion with Jacob yesterday, we both agreed that we would remove it for the moment and the user can simply do `clusterer.learn_one().predict_one()` as an equivalence to this function.